File: backend/app/schemas/surveys.py
"""Schemas for the WTP intercept survey."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


class SurveyCreate(BaseModel):
    venue_label: str = Field(min_length=1, max_length=120)
    observed_wait_mins: Optional[int] = Field(default=None, ge=0, le=600)
    party_size: int = Field(ge=1, le=20)
    respondent: Literal["tourist", "local"]
    would_skip: bool
    max_fee_yen: Optional[int] = Field(default=None, ge=0, le=50_000)  # legacy, no longer collected
    offered_price_yen: Optional[int] = Field(default=None, ge=100, le=50_000)
    perceived_wait_mins: Optional[int] = Field(default=None, ge=0, le=600)
    stated_max_wait_mins: Optional[int] = Field(default=None, ge=0, le=600)
    time_pressure: Optional[Literal["hurry", "normal", "relaxed"]] = None
    first_visit: Optional[bool] = None
    reason: Optional[str] = Field(default=None, max_length=80)
    notes: Optional[str] = Field(default=None, max_length=500)


class SurveyRead(SurveyCreate):
    id: uuid.UUID
    restaurant_id: uuid.UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/lib/surveys.ts
// WTP survey client.

import { api } from "./api";

export interface SurveyCreate {
  venue_label: string;
  observed_wait_mins?: number | null;
  party_size: number;
  respondent: "tourist" | "local";
  would_skip: boolean;
  max_fee_yen?: number | null; // legacy, no longer collected
  offered_price_yen?: number | null;
  perceived_wait_mins?: number | null;
  stated_max_wait_mins?: number | null;
  time_pressure?: "hurry" | "normal" | "relaxed" | null;
  first_visit?: boolean | null;
  reason?: string | null;
  notes?: string | null;
}

export interface SurveyRead extends SurveyCreate {
  id: string;
  restaurant_id: string | null;
  created_at: string;
}

export interface SurveySummary {
  total: number;
  overall_yes_rate: number | null;
  median_stated_max_wait: number | null;
  by_price: { price: number; n: number; yes: number; yes_rate: number | null }[];
  by_respondent: Record<string, { n: number; yes_rate: number | null }>;
  by_venue: { venue: string; n: number; yes_rate: number | null }[];
}

export const surveysApi = {
  summary: (token: string) => api<SurveySummary>("/api/surveys/summary", { token }),
  create: (token: string, body: SurveyCreate) =>
    api<SurveyRead>("/api/surveys", { method: "POST", body, token }),
  list: (token: string, limit = 200) =>
    api<SurveyRead[]>(`/api/surveys?limit=${limit}`, { token }),
};
